Add "security" commands to api CLI

danilopolani opened the pull request:
This PR implements two new CLI commands. I thought they could be useful especially when cloning a project and, of course, you have to setup your own `.env` file or when deploying on production  and need to setup the environment variables.

### Key generation

It generates the `KEY` variable. You can additionally use the `--show` flag to avoid the `.env` overwrite and just display it in the terminal.

```bash
$ npx directus security key:generate
# App key set correctly.

$ npx directus security key:generate --show
# Your new app key: 9a30bf49-946c-430c-83fc-eaefce5673c3
```

### Secret generation

It generates the `SECRET` variable. You can additionally use the `--show` flag to avoid the `.env` overwrite and just display it in the terminal.

```bash
$ npx directus security secret:generate
# App secret set correctly.

$ npx directus security secret:generate --show
# Your new app secret: Ypzy1c2O3tcvpP_1c7uT4Dhl-liOub8S
```

Both commands will fail if no `.env` file is found in the cwd and the flag `--show` is omitted:

![image](https://user-images.githubusercontent.com/6277291/140049779-8adc2a87-9aa0-4f8f-b04e-f01036396930.png)

----

**Notes**: I named them like this because I'm used to Laravel `key:generate --show` command, but let me know if you'd prefer another name or if we should group them under a `security` group, like commands under the `users` group etc.

danilopolani committed:
add command to generate app key

danilopolani committed:
add command to generate secret

danilopolani committed:
Merge branch 'directus:main' into feat/cli-security-commands

benhaynes commented:
We can wait until this change is approved, but we'll need to make sure docs are also added as part of this PR.

danilopolani commented:
> We can wait until this change is approved, but we'll need to make sure docs are also added as part of this PR.

Right! I totally forgot the docs, sorry, I'm going to work on it later today

danilopolani committed:
group security commands under the security group

danilopolani committed:
add new security commands to cli docs

danilopolani committed:
add disclaimer on deploy secret generation

danilopolani committed:
Merge branch 'feat/cli-security-commands' of github.com:danilopolani/directus into feat/cli-security-commands

danilopolani committed:
remove duplicate hubspot embed code

danilopolani commented:
A few updates:

1. The two commands are now grouped under the `security` group, like databases commands and so on;
2. I've updated the documentation, feel free to edit any paragraph since english it's not my native language 😄 
3. I've added an alert in the installation page about the generation of a fresh token when deploying, let me know if you'd prefer to remove it:

![image](https://user-images.githubusercontent.com/6277291/140101807-87a91b53-f5ac-4fa1-898e-d67827a512a9.png)

----

Oh, and the HubSpot embed code was - for some reason - embedded two times in some page and I had to remove the duplicates because Vue was complaining about multiple root elements

rijkvanzanten commented on api/src/cli/commands/security/key.ts:
```diff
@@ -0,0 +1,30 @@
+	}
+
+	const rootPath = process.cwd();
+	const envPath = pathJoin(rootPath, '.env');
```
The path where the config file resides is user configurable, so we can't rely on it being the root + '.env'

rijkvanzanten commented on api/src/cli/index.ts:
```diff
@@ -30,6 +32,19 @@ export async function createCli(): Promise<Command> {
+	securityCommand
+		.command('key:generate')
+		.description('Generate the app key')
+		.option('--show', 'Show the key without updating the .env file')
```
I think the only thing we can do is make this the default, and remove the writing to .env altogether

danilopolani commented on api/src/cli/commands/security/key.ts:
```diff
@@ -0,0 +1,30 @@
+		process.exit(1);
+	}
+
+	const envFileContent = readFileSync(envPath)
+		.toString()
+		.replace(/^KEY=(.*)$/gm, `KEY="${key}"`);
```
Uhm, for quotes/not-quotes it's not really a problem, here we just replace the line starting with `KEY=`, so it matches `KEY="whatever"` and `KEY=whatever`. About YAML/JSON maybe we can detect it (?) or just let user choose the format with `--yaml` or `--json`? For `docker-compose` they should use `--show`...

...Weeeeell, ok, I guess it's better to just leave the `--show` behaviour as default haha

rijkvanzanten replied:
> here we just replace the line starting with KEY=, so it matches KEY="whatever" and KEY=whatever

True, but right now you're replacing the whole line with the quotes version 🙂 I just recall there was at least one environment where the difference in quotes mattered, it might've been Docker though though, not the .env file. 

We could make it work for JSON/YAML, but that doesn't solve it for the function-based config setup, which allows the user to dynamically generate the config in an async function on start 🤔

danilopolani replied:
Yeah, you convinced me 👀 I'm going to push the edit soon

danilopolani committed:
remove .env overwrite feature to security commands

danilopolani committed:
update the cli docs for security commands

danilopolani commented:
Ok, at the end I removed the feature to automatically update the `.env` with generated values because it would be a pain to develop and maintain support for YAML, JSON and every variation of them

danilopolani committed:
Merge branch 'directus:main' into feat/cli-security-commands

rijkvanzanten commented on api/src/cli/commands/security/key.ts:
```diff
@@ -0,0 +1,8 @@
+export default async function generateKey(): Promise<void> {
+	const key = uuidv4();
+
+	process.stdout.write(`Your new app key: ${key}\n`);
```
Lets return the value immediately, instead of prefixing it with the message. That'll allow CI to use it as well (where you can pipe the output into something else)

danilopolani replied:
I totally forgot about this PR haha. Should we still return a new line after the value?

rijkvanzanten replied:
> Should we still return a new line after the value?

Nah, lets just return the raw value and nothing else. That way it's easiest for external processes to use the generated information 🙂

danilopolani replied:
Changed!

rijkvanzanten replied:
Yup!

danilopolani committed:
return key generation directly in output

danilopolani committed:
Merge branch 'feat/cli-security-commands' of github.com:danilopolani/directus into feat/cli-security-commands

danilopolani committed:
merge from main

danilopolani committed:
restore package-lock.json

rijkvanzanten committed:
Write secret directly
